python/tank/descriptor/bundle_cache_usage/logger.py

```diff
@@ -0,0 +1,281 @@
+                # or requesting end of thread.
+                self._queued_signal.clear()
+
+        except Exception as e:
+            self.__log_error(e, "Unexpected exception in the worker run() method")
+
+        tmp_log.debug_worker_threading("terminated (%d tasks remaining)" % self.pending_count)
+
+    #
+    # Protected methods
+    # Can run from either threading contextes
+    #
+    ###########################################################################
+
+    def _check_for_queued_errors(self):
+        while not self._errors.empty():
+            exception, message = self._errors.get()
+            log.error("%s : %s" % (message, exception))
+
+    def _queue_task(self, method, *args, **kwargs):
+        """
+        Add a task to the worker thread queue.
+
+        .. note:: Called from main thread
+
+        :param method: method to call
+        :param args: arguments to pass to the method
+        :param kwargs: named arguments to pass to the method
+        """
+        self._check_for_queued_errors()
+
+        with self._member_lock:
+            self._pending_count += 1
+
+        self._tasks.put((method, args, kwargs))
+        self._tasks.task_done()
+        self._queued_signal.set()
+
+    #
+    # Public methods & properties
+    # Can run from either threading contextes
+    #
+    ###########################################################################
+
+    @property
+    def bundle_cache_root(self):
+        """
+        Returns the bundle cache path initially supplied in constructor.
+
+        :return: A str path
+        """
+        return self._bundle_cache_root
+
+    @property
+    def completed_count(self):
+        """
+        Returns how many tasks have been completed by the worker thread.
+
+        .. note:: Indicative only, with the worker thread running,
+        by the time the next instruction is executed the value might
+        be already different.
+
+        :return: an integer of the completed task count
+        """
+        with self._member_lock:
+            return self._completed_count
+
+    @classmethod
+    def log_usage(cls, bundle_path):
+        """
+        Queue a bundle cache usage database update.
+
+        :param bundle_path: a str bundle path
+        """
+        with cls.__singleton_lock:
+            if cls.__singleton_instance:
+                cls.__singleton_instance._queue_task(cls.__singleton_instance.__log_usage, bundle_path)
+
+    @property
+    def pending_count(self):
+        """
+        Returns how many tasks are in the queue, still pending.
+
+        .. note:: Indicative only, with the worker thread running,
+        by the time the next instruction is executed the value might
+        be already different.
+
+        :return: an integer of the currently still queued tasks count
+        """
+        with self._member_lock:
+            return self._pending_count
+
+    def stop(self, timeout=DEFAULT_STOP_TIMEOUT):
```

Comment:
if we just use the queue, can this just be

```
def stop(self):
    self._tasks.put((QUIT))
```

Feels like we have a lot of complicated threading constructs. In my experience, threading is where you go to die. I would like to have a minimal shared state between threads and i would love if we can shrink this code back to the bare minimum for maintenance reasons.